### ISSUE TITLE
[JENKINS-47908] After upgrade to v2.0, the active choices updates randomly in the browser

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/active-choices-plugin#699]
<p>Hello,</p>

<p>I used Active choices parameters version 1.5.2 in a production environment since yesterday.<br/>
 I updated the server and got version 2.0, whitelisted the necessary java methods and then tested the new behaviour of the "build with parameters" page.</p>

<p>When I modify a parameter's value, the other parameters updates only randomly, or partially (some are up to date, while the other not). The page sometimes blinked while updating, in which case more parameters were up to date (seems like an async bug to me).</p>

<p>My setup is :<br/>
 Action : build_partial/build_full<br/>
 Part : part1/part2:selected/part3/full   => full is display when build_full is chosen and replaces the other choices<br/>
 Subpart : subpart1/subpart2/subpart3    => subparts are displayes only for part2<br/>
 Other fields that are empty if action is build_full.</p>

<p>Then I rolled back to 1.5.2 and everything worked fine again.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kiruahxh">kiruahxh</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908" target="_blank">After upgrade to v2.0, the active choices updates randomly in the browser</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow">kinow</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Blocker
<li><b>component(s)</b>: active-choices-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-11-29T20:40:08+00:00
<li><b>votes</b>: 8
<li><b>watchers</b>: 13
<li><b>imported</b>: 2025-12-08
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Hello,</p>

<p>I used Active choices parameters version 1.5.2 in a production environment since yesterday.
 I updated the server and got version 2.0, whitelisted the necessary java methods and then tested the new behaviour of the "build with parameters" page.</p>

<p>When I modify a parameter's value, the other parameters updates only randomly, or partially (some are up to date, while the other not). The page sometimes blinked while updating, in which case more parameters were up to date (seems like an async bug to me).</p>

<p>My setup is :
 Action : build_partial/build_full
 Part : part1/part2:selected/part3/full   =&gt; full is display when build_full is chosen and replaces the other choices
 Subpart : subpart1/subpart2/subpart3    =&gt; subparts are displayes only for part2
 Other fields that are empty if action is build_full.</p>

<p>Then I rolled back to 1.5.2 and everything worked fine again.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Version 1.5.2<br/>
Firefox ESR
```
</details>
<details><summary><i>7 attachments</i></summary>

- [image-2017-11-21-15-09-31-378.png](https://issues.jenkins.io/secure/attachment/40446/image-2017-11-21-15-09-31-378.png)
  > ![image-2017-11-21-15-09-31-378.png](https://issues.jenkins.io/secure/attachment/40446/image-2017-11-21-15-09-31-378.png)
- [image-2017-11-21-15-10-20-995.png](https://issues.jenkins.io/secure/attachment/40447/image-2017-11-21-15-10-20-995.png)
  > ![image-2017-11-21-15-10-20-995.png](https://issues.jenkins.io/secure/attachment/40447/image-2017-11-21-15-10-20-995.png)
- [image-2017-11-21-15-11-15-525.png](https://issues.jenkins.io/secure/attachment/40448/image-2017-11-21-15-11-15-525.png)
  > ![image-2017-11-21-15-11-15-525.png](https://issues.jenkins.io/secure/attachment/40448/image-2017-11-21-15-11-15-525.png)
- [image-2017-11-21-15-12-28-617.png](https://issues.jenkins.io/secure/attachment/40449/image-2017-11-21-15-12-28-617.png)
  > ![image-2017-11-21-15-12-28-617.png](https://issues.jenkins.io/secure/attachment/40449/image-2017-11-21-15-12-28-617.png)
- [image-2017-11-21-15-24-06-520.png](https://issues.jenkins.io/secure/attachment/40450/image-2017-11-21-15-24-06-520.png)
  > ![image-2017-11-21-15-24-06-520.png](https://issues.jenkins.io/secure/attachment/40450/image-2017-11-21-15-24-06-520.png)
- [image-2017-12-08-11-56-10-561.png](https://issues.jenkins.io/secure/attachment/40626/image-2017-12-08-11-56-10-561.png)
  > ![image-2017-12-08-11-56-10-561.png](https://issues.jenkins.io/secure/attachment/40626/image-2017-12-08-11-56-10-561.png)
- [uno-choice.hpi](https://issues.jenkins.io/secure/attachment/40615/uno-choice.hpi)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [CloudBees Internal OSS-2585](https://cloudbees.atlassian.net/browse/OSS-2585)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>datom:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=319649&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319649">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hey!</p>

<p>We have exactly the same problem since we updated to 2.0.</p>

<p>For us it looks like this only happens if you use multiple 'Active Choices Reactive Reference Parameter' referencing the same Parameter ("Referenced parameters").</p>

<p>BR,</p>

<p>Thomas</p>
</pre>
</details>

</ul>
</details>

<p>Hey!</p>

<p>We have exactly the same problem since we updated to 2.0.</p>

<p>For us it looks like this only happens if you use multiple 'Active Choices Reactive Reference Parameter' referencing the same Parameter ("Referenced parameters").</p>

<p>BR,</p>

<p>Thomas</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>heitec_plc3:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=319664&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319664">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello,</p>

<p>also we have the same problem after we updated to version 2.0.</p>

<p>We use Groovy-Script. Our workaround at the moment we use plug-in version 1.4 at the latest Jenkins-version 2.89. So far it works but should not be the final solution to us. If we may get a fixed plug-in we would be really glad.</p>

<p>BR,</p>

<p>Frank</p>
</pre>
</details>

</ul>
</details>

<p>Hello,</p>

<p>also we have the same problem after we updated to version 2.0.</p>

<p>We use Groovy-Script. Our workaround at the moment we use plug-in version 1.4 at the latest Jenkins-version 2.89. So far it works but should not be the final solution to us. If we may get a fixed plug-in we would be really glad.</p>

<p>BR,</p>

<p>Frank</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>carek:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=319757&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319757">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>In my case I have a drop down box which change should trigger update of two read-only text fields. Fields are sometimes refreshed properly (very rare case), sometimes only one of them is refreshed, sometimes nothing happens at all. I am not sure what kind of code is underneath, but it looks like event swallowing or something. In order to get correct values I always need to change value of drop down box multiple times. Since I use it in deployment jobs, it happens quite often that our environments get wiped out because wrong addresses are set <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>please fix this behavior asap <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>In my case I have a drop down box which change should trigger update of two read-only text fields. Fields are sometimes refreshed properly (very rare case), sometimes only one of them is refreshed, sometimes nothing happens at all. I am not sure what kind of code is underneath, but it looks like event swallowing or something. In order to get correct values I always need to change value of drop down box multiple times. Since I use it in deployment jobs, it happens quite often that our environments get wiped out because wrong addresses are set <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>please fix this behavior asap <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>myrsh:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=319760&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319760">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We see the same behavior at our end. I understand that scripts can now run in parallel. Thus they have access to a threadpool.</p>

<p>I would request configurable Plugin attribute like "threadpool.size" which will let users configure the behavior.</p>

<p>Example. If we need serial behavior then we will keep threadpool size = 1. </p>
</pre>
</details>

</ul>
</details>

<p>We see the same behavior at our end. I understand that scripts can now run in parallel. Thus they have access to a threadpool.</p>

<p>I would request configurable Plugin attribute like "threadpool.size" which will let users configure the behavior.</p>

<p>Example. If we need serial behavior then we will keep threadpool size = 1. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=319924&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319924">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for reporting the issue, and thanks all for chiming in and confirming you are also able to reproduce. This information is extremely useful while working on tickets for a new release.</p>

<p>We have a new BioUno meeting in a couple days, and will discuss and prioritize issues. This one will probably be fixed in the next development cycle. There is another pull request pending, that just needs a bit of testing, but the code has been already reviewed.</p>

<p>The plug-in is maintained by volunteers during their spare time, with occasional contributions by private companies that were paid to work on issues. So can't promise a schedule.</p>

<p>But stay tuned, and we are always happy to review pull requests. What also helps is a specific version of Jenkins + specific version of the plug-in (clear in this ticket) + a config.xml with a minimal example to reproduce the bug. That saves us hours sometimes, depending on the ticket.</p>

<p>Thanks once again.</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for reporting the issue, and thanks all for chiming in and confirming you are also able to reproduce. This information is extremely useful while working on tickets for a new release.</p>

<p>We have a new BioUno meeting in a couple days, and will discuss and prioritize issues. This one will probably be fixed in the next development cycle. There is another pull request pending, that just needs a bit of testing, but the code has been already reviewed.</p>

<p>The plug-in is maintained by volunteers during their spare time, with occasional contributions by private companies that were paid to work on issues. So can't promise a schedule.</p>

<p>But stay tuned, and we are always happy to review pull requests. What also helps is a specific version of Jenkins + specific version of the plug-in (clear in this ticket) + a config.xml with a minimal example to reproduce the bug. That saves us hours sometimes, depending on the ticket.</p>

<p>Thanks once again.</p>

<p>Bruno</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>klimas7:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320423&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320423">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a></p>

<p>This issue is results of the change which you done in <a href="https://issues.jenkins-ci.org/browse/JENKINS-42685" class="external-link" target="_blank" rel="nofollow noopener">JENKINS-42685</a></p>

<p>This is very simple test configuration </p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40446/40446_image-2017-11-21-15-09-31-378.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40447/40447_image-2017-11-21-15-10-20-995.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40448/40448_image-2017-11-21-15-11-15-525.png" style="border: 0px solid black" /></span></p>

<p>You get resault:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40449/40449_image-2017-11-21-15-12-28-617.png" style="border: 0px solid black" /></span></p>

<p>Web console output:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Updating cascade of parameter [param_b] ...
Values retrieved from Referenced Parameters: param_a=a
Calling Java server code to update HTML elements...
Updating cascade of parameter [param_c] ...
Values retrieved from Referenced Parameters: param_a=a__LESEP__param_b=
Calling Java server code to update HTML elements...
Values returned from server: [[<span class="code-quote">"a1"</span>,<span class="code-quote">"a2"</span>,<span class="code-quote">"a3"</span>],[<span class="code-quote">"a1"</span>,<span class="code-quote">"a2"</span>,<span class="code-quote">"a3"</span>]]
Values returned from server: [[],[]]
</pre>
</div></div>
<p> </p>

<p><b>Before change <a href="https://issues.jenkins.io/browse/JENKINS-42685" title="Remove custom stapler proxy as we can now use assync requests" class="issue-link" data-issue-key="JENKINS-42685"><del>JENKINS-42685</del></a> that worked like that:</b></p>

<p> </p>

<p>*Updating cascade of parameter <span class="error">&#91;param_b&#93;</span> ...
Values retrieved from Referenced Parameters: param_a=a
<span class="error">&#91;Deprecation&#93;</span> Synchronous XMLHttpRequest ... <a href="https://xhr.spec.whatwg.org/" class="external-link" target="_blank" rel="nofollow noopener">https://xhr.spec.whatwg.org/</a>.
Calling Java server code to update HTML elements...
Values returned from server: [<span class="error">&#91;&quot;a1&quot;,&quot;a2&quot;,&quot;a3&quot;&#93;</span>,<span class="error">&#91;&quot;a1&quot;,&quot;a2&quot;,&quot;a3&quot;&#93;</span>]
Updating cascade of parameter <span class="error">&#91;param_c&#93;</span> ...
Values retrieved from Referenced Parameters: param_a=a_<em>LESEP</em>_param_b=a1
Calling Java server code to update HTML elements...
Values returned from server: [<span class="error">&#91;&quot;a11&quot;,&quot;a12&quot;,&quot;a13&quot;&#93;</span>,<span class="error">&#91;&quot;a11&quot;,&quot;a12&quot;,&quot;a13&quot;&#93;</span>]*</p>

<p><b><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40450/40450_image-2017-11-21-15-24-06-520.png" style="border: 0px solid black" /></span></b></p>

<p> </p>

<p> </p>

<p> I added test groovy scripts for you</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-comment">//param_a
</span><span class="code-keyword">return</span> [<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>,<span class="code-quote">'c'</span>]

<span class="code-comment">//param_b depends on param_a
</span>def map = [<span class="code-quote">'a'</span>:[<span class="code-quote">'a1'</span>,<span class="code-quote">'a2'</span>,<span class="code-quote">'a3'</span>], <span class="code-quote">'b'</span>:[<span class="code-quote">'b1'</span>,<span class="code-quote">'b2'</span>,<span class="code-quote">'b3'</span>], <span class="code-quote">'c'</span>:[<span class="code-quote">'c1'</span>,<span class="code-quote">'c2'</span>,<span class="code-quote">'c3'</span>]]
<span class="code-keyword">return</span> map[param_a]

<span class="code-comment">//param_c depends on param_a, param_b
</span>def map_a = [<span class="code-quote">'a1'</span>:[<span class="code-quote">'a11'</span>,<span class="code-quote">'a12'</span>,<span class="code-quote">'a13'</span>],<span class="code-quote">'a2'</span>:[<span class="code-quote">'a21'</span>,<span class="code-quote">'a22'</span>,<span class="code-quote">'a23'</span>],<span class="code-quote">'a3'</span>:[<span class="code-quote">'a31'</span>,<span class="code-quote">'a32'</span>,<span class="code-quote">'a33'</span>]]
def map_b = [<span class="code-quote">'b1'</span>:[<span class="code-quote">'b11'</span>,<span class="code-quote">'b12'</span>,<span class="code-quote">'b13'</span>],<span class="code-quote">'b2'</span>:[<span class="code-quote">'b21'</span>,<span class="code-quote">'b22'</span>,<span class="code-quote">'b23'</span>],<span class="code-quote">'b3'</span>:[<span class="code-quote">'b31'</span>,<span class="code-quote">'b32'</span>,<span class="code-quote">'b33'</span>]]
def map_c = [<span class="code-quote">'c1'</span>:[<span class="code-quote">'c11'</span>,<span class="code-quote">'c12'</span>,<span class="code-quote">'c13'</span>],<span class="code-quote">'c2'</span>:[<span class="code-quote">'c21'</span>,<span class="code-quote">'c22'</span>,<span class="code-quote">'c23'</span>],<span class="code-quote">'c3'</span>:[<span class="code-quote">'c31'</span>,<span class="code-quote">'c32'</span>,<span class="code-quote">'c33'</span>]]
def master_map = [<span class="code-quote">'a'</span>:map_a, <span class="code-quote">'b'</span>:map_b, <span class="code-quote">'c'</span>:map_c] 
<span class="code-keyword">return</span> master_map[param_a][param_b]</pre>
</div></div>
<p> </p>

<p> </p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a></p>

<p>This issue is results of the change which you done in <a href="https://issues.jenkins-ci.org/browse/JENKINS-42685" class="external-link" target="_blank" rel="nofollow noopener">JENKINS-42685</a></p>

<p>This is very simple test configuration </p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40446/image-2017-11-21-15-09-31-378.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/40447/image-2017-11-21-15-10-20-995.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/40448/image-2017-11-21-15-11-15-525.png" style="border: 0px solid black" /></span></p>

<p>You get resault:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40449/image-2017-11-21-15-12-28-617.png" style="border: 0px solid black" /></span></p>

<p>Web console output:</p>

<pre>
Updating cascade of parameter [param_b] ...
Values retrieved from Referenced Parameters: param_a=a
Calling Java server code to update HTML elements...
Updating cascade of parameter [param_c] ...
Values retrieved from Referenced Parameters: param_a=a__LESEP__param_b=
Calling Java server code to update HTML elements...
Values returned from server: [[<span class="code-quote">"a1"</span>,<span class="code-quote">"a2"</span>,<span class="code-quote">"a3"</span>],[<span class="code-quote">"a1"</span>,<span class="code-quote">"a2"</span>,<span class="code-quote">"a3"</span>]]
Values returned from server: [[],[]]
</pre>
<p> </p>

<p><b>Before change <a href="https://issue-redirect.jenkins.io/issue/42685" title="Remove custom stapler proxy as we can now use assync requests" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/42685">JENKINS-42685</a>"><del>JENKINS-42685</del></a> that worked like that:</b></p>

<p> </p>

<p>*Updating cascade of parameter <span class="error">&#91;param_b&#93;</span> ...<br/>
Values retrieved from Referenced Parameters: param_a=a<br/>
<span class="error">&#91;Deprecation&#93;</span> Synchronous XMLHttpRequest ... <a href="https://xhr.spec.whatwg.org/" class="external-link" target="_blank" rel="nofollow noopener">https://xhr.spec.whatwg.org/</a>.<br/>
Calling Java server code to update HTML elements...<br/>
Values returned from server: [<span class="error">&#91;"a1","a2","a3"&#93;</span>,<span class="error">&#91;"a1","a2","a3"&#93;</span>]<br/>
Updating cascade of parameter <span class="error">&#91;param_c&#93;</span> ...<br/>
Values retrieved from Referenced Parameters: param_a=a_<em>LESEP</em>_param_b=a1<br/>
Calling Java server code to update HTML elements...<br/>
Values returned from server: [<span class="error">&#91;"a11","a12","a13"&#93;</span>,<span class="error">&#91;"a11","a12","a13"&#93;</span>]*</p>

<p><b><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/40450/image-2017-11-21-15-24-06-520.png" style="border: 0px solid black" /></span></b></p>

<p> </p>

<p> </p>

<p> I added test groovy scripts for you</p>

<pre>
<span class="code-comment">//param_a
</span><span class="code-keyword">return</span> [<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>,<span class="code-quote">'c'</span>]

<span class="code-comment">//param_b depends on param_a
</span>def map = [<span class="code-quote">'a'</span>:[<span class="code-quote">'a1'</span>,<span class="code-quote">'a2'</span>,<span class="code-quote">'a3'</span>], <span class="code-quote">'b'</span>:[<span class="code-quote">'b1'</span>,<span class="code-quote">'b2'</span>,<span class="code-quote">'b3'</span>], <span class="code-quote">'c'</span>:[<span class="code-quote">'c1'</span>,<span class="code-quote">'c2'</span>,<span class="code-quote">'c3'</span>]]
<span class="code-keyword">return</span> map[param_a]

<span class="code-comment">//param_c depends on param_a, param_b
</span>def map_a = [<span class="code-quote">'a1'</span>:[<span class="code-quote">'a11'</span>,<span class="code-quote">'a12'</span>,<span class="code-quote">'a13'</span>],<span class="code-quote">'a2'</span>:[<span class="code-quote">'a21'</span>,<span class="code-quote">'a22'</span>,<span class="code-quote">'a23'</span>],<span class="code-quote">'a3'</span>:[<span class="code-quote">'a31'</span>,<span class="code-quote">'a32'</span>,<span class="code-quote">'a33'</span>]]
def map_b = [<span class="code-quote">'b1'</span>:[<span class="code-quote">'b11'</span>,<span class="code-quote">'b12'</span>,<span class="code-quote">'b13'</span>],<span class="code-quote">'b2'</span>:[<span class="code-quote">'b21'</span>,<span class="code-quote">'b22'</span>,<span class="code-quote">'b23'</span>],<span class="code-quote">'b3'</span>:[<span class="code-quote">'b31'</span>,<span class="code-quote">'b32'</span>,<span class="code-quote">'b33'</span>]]
def map_c = [<span class="code-quote">'c1'</span>:[<span class="code-quote">'c11'</span>,<span class="code-quote">'c12'</span>,<span class="code-quote">'c13'</span>],<span class="code-quote">'c2'</span>:[<span class="code-quote">'c21'</span>,<span class="code-quote">'c22'</span>,<span class="code-quote">'c23'</span>],<span class="code-quote">'c3'</span>:[<span class="code-quote">'c31'</span>,<span class="code-quote">'c32'</span>,<span class="code-quote">'c33'</span>]]
def master_map = [<span class="code-quote">'a'</span>:map_a, <span class="code-quote">'b'</span>:map_b, <span class="code-quote">'c'</span>:map_c] 
<span class="code-keyword">return</span> master_map[param_a][param_b]</pre>
<p> </p>

<p> </p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320513&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320513">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=klimas7" class="user-hover" rel="klimas7">klimas7</a> thanks for spending time to look into this. What a detective work! Having a change to compare will make it way easier to reproduce, test, and fix.</p>

<p>I have a deadline on December 3rd for another obligation, but am trying to find some spare time before that to prepare a new release with this issue fixed. Otherwise I will look into this issue after that weekend.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=klimas7" class="user-hover" rel="klimas7">klimas7</a> thanks for spending time to look into this. What a detective work! Having a change to compare will make it way easier to reproduce, test, and fix.</p>

<p>I have a deadline on December 3rd for another obligation, but am trying to find some spare time before that to prepare a new release with this issue fixed. Otherwise I will look into this issue after that weekend.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320831&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320831">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Bruno P. Kinoshita
Path:
 src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/index.jelly
 src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/index.jelly
 src/main/resources/org/biouno/unochoice/stapler/unochoice/unochoice.js
<a href="http://jenkins-ci.org/commit/active-choices-plugin/24a60d664ec37c9de3d64de5de5c330d98c230e4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/active-choices-plugin/24a60d664ec37c9de3d64de5de5c330d98c230e4</a>
Log:
  <span class="error">&#91;FIXED JENKINS-47908&#93;</span>: Revert "<a href="https://issues.jenkins.io/browse/JENKINS-42685" title="Remove custom stapler proxy as we can now use assync requests" class="issue-link" data-issue-key="JENKINS-42685"><del>JENKINS-42685</del></a> Remove stapler proxy as we can now use assync requests"</p>

<p>This reverts commit 8ccd7fcdec84e3cb6272115662632fa2542ce945.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Bruno P. Kinoshita<br/>
Path:<br/>
 src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/index.jelly<br/>
 src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/index.jelly<br/>
 src/main/resources/org/biouno/unochoice/stapler/unochoice/unochoice.js<br/>
<a href="http://jenkins-ci.org/commit/active-choices-plugin/24a60d664ec37c9de3d64de5de5c330d98c230e4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/active-choices-plugin/24a60d664ec37c9de3d64de5de5c330d98c230e4</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED <a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/47908">JENKINS-47908</a>&#93;</span>: Revert "<a href="https://issue-redirect.jenkins.io/issue/42685" title="Remove custom stapler proxy as we can now use assync requests" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/42685">JENKINS-42685</a>"><del>JENKINS-42685</del></a> Remove stapler proxy as we can now use assync requests"</p>

<p>This reverts commit 8ccd7fcdec84e3cb6272115662632fa2542ce945.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320834&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320834">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Reverted the commit. Going to work on a few other tickets, and then ask others to, please, review and test a release candidate.</p>

<p> </p>

<p>Cheers</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Reverted the commit. Going to work on a few other tickets, and then ask others to, please, review and test a release candidate.</p>

<p> </p>

<p>Cheers</p>

<p>Bruno</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320898&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320898">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks Bruno - I'll be happy to test out any RC since we were also experiencing issues with the update.  Any chance <a href="https://issues.jenkins-ci.org/browse/JENKINS-45507" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-45507</a> could be included (if it helps solve the update ordering issue)?</p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>Thanks Bruno - I'll be happy to test out any RC since we were also experiencing issues with the update.  Any chance <a href="https://issues.jenkins-ci.org/browse/JENKINS-45507" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-45507</a> could be included (if it helps solve the update ordering issue)?</p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320953&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320953">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'd prefer to release a 2.1 with this fix and the other pull request. Both changes are done, and ready to be tested, I just want to see if there's an easy ticket to include in the release before asking for help testing.</p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-45507" title="Evaluate parameters once" class="issue-link" data-issue-key="JENKINS-45507">JENKINS-45507</a> is in our roadmap for the next releases, but given we just caused a regression, I'd prefer to wait a bit before risking a change like this. I'll probably create a branch for <a href="https://issues.jenkins.io/browse/JENKINS-45507" title="Evaluate parameters once" class="issue-link" data-issue-key="JENKINS-45507">JENKINS-45507</a>, and then will need help to test that as well. Will ping you once we start working on a 2.2 release (December/January). How does that sound?</p>
</pre>
</details>

</ul>
</details>

<p>I'd prefer to release a 2.1 with this fix and the other pull request. Both changes are done, and ready to be tested, I just want to see if there's an easy ticket to include in the release before asking for help testing.</p>

<p><a href="https://issue-redirect.jenkins.io/issue/45507" title="Evaluate parameters once" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/45507">JENKINS-45507</a>">JENKINS-45507</a> is in our roadmap for the next releases, but given we just caused a regression, I'd prefer to wait a bit before risking a change like this. I'll probably create a branch for <a href="https://issue-redirect.jenkins.io/issue/45507" title="Evaluate parameters once" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/45507">JENKINS-45507</a>">JENKINS-45507</a>, and then will need help to test that as well. Will ping you once we start working on a 2.2 release (December/January). How does that sound?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=320984&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320984">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yup, no worries - although I super look forward to <a href="https://issues.jenkins.io/browse/JENKINS-45507" title="Evaluate parameters once" class="issue-link" data-issue-key="JENKINS-45507">JENKINS-45507</a>, I understand you don't want to introduce feature creep, and want to get these few updates working first. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  </p>

<p>Bill</p>
</pre>
</details>

</ul>
</details>

<p>Yup, no worries - although I super look forward to <a href="https://issue-redirect.jenkins.io/issue/45507" title="Evaluate parameters once" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/45507">JENKINS-45507</a>">JENKINS-45507</a>, I understand you don't want to introduce feature creep, and want to get these few updates working first. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  </p>

<p>Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=321280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-321280">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Marking as fixed for now. We will start testing soon. If anyone find an issue, this can be re-opened.</p>
</pre>
</details>

</ul>
</details>

<p>Marking as fixed for now. We will start testing soon. If anyone find an issue, this can be re-opened.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=322194&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322194">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ioannis" class="user-hover" rel="ioannis">ioannis</a>, attaching uno-choice.hpi (de4ab1b29fd32361f447cbd153900bdb). Ready to be tested!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ioannis" class="user-hover" rel="ioannis">ioannis</a>, attaching uno-choice.hpi (de4ab1b29fd32361f447cbd153900bdb). Ready to be tested!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>klimas7:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=322195&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322195">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Bruno,</p>

<p>I tested this issue, and its ok. Plugin work fine.</p>

<p>Boguslaw</p>
</pre>
</details>

</ul>
</details>

<p>Hi Bruno,</p>

<p>I tested this issue, and its ok. Plugin work fine.</p>

<p>Boguslaw</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=322196&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322196">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Boguslaw!</p>

<p>One of the fastest replies I ever got in all my time in open source <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> thanks for the feedback!!!</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Hi Boguslaw!</p>

<p>One of the fastest replies I ever got in all my time in open source <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> thanks for the feedback!!!</p>

<p>Bruno</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> ioannis:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=322264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322264">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a> Thanks for making the hpi available for testing!</p>

<p>The new plugin requires Jenkins v 2.46.3 or later. That is due to the new security framework.</p>

<p>After I upgraded my Jenkins instance to v 2.93 (latest at the moment) and Active Choices v2.0 (patch) the parameter synchonization issue seems fixed. THANKS!</p>

<p> </p>

<p><b>However, HTML in parameter descriptions is no longer rendered</b>.</p>

<p>Here is an example</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40626/40626_image-2017-12-08-11-56-10-561.png" style="border: 0px solid black" /></span></p>

<p>Not sure whether this is fixable in Active Choices or some other Jenkins security setting. But it was fixed  for a related parameter type.</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-42903" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-42903</a></p>

<p>I rely a lot on HTML in descriptions (even dynamically changing them with AC parameters)  and this is a blocker for several of my Jenkins projects. </p>

<p>I will need to wait for a patch on this before I can deploy to production. I will also open another JIRA issue.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a> Thanks for making the hpi available for testing!</p>

<p>The new plugin requires Jenkins v 2.46.3 or later. That is due to the new security framework.</p>

<p>After I upgraded my Jenkins instance to v 2.93 (latest at the moment) and Active Choices v2.0 (patch) the parameter synchonization issue seems fixed. THANKS!</p>

<p> </p>

<p><b>However, HTML in parameter descriptions is no longer rendered</b>.</p>

<p>Here is an example</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40626/image-2017-12-08-11-56-10-561.png" style="border: 0px solid black" /></span></p>

<p>Not sure whether this is fixable in Active Choices or some other Jenkins security setting. But it was fixed  for a related parameter type.</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-42903" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-42903</a></p>

<p>I rely a lot on HTML in descriptions (even dynamically changing them with AC parameters)  and this is a blocker for several of my Jenkins projects. </p>

<p>I will need to wait for a patch on this before I can deploy to production. I will also open another JIRA issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-47908?focusedId=322288&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322288">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Ioannis!</p>

<p>Thanks for testing it so quickly. Perhaps we can work on a patch for this one too? First I will investigate if we can enable the HTML for those fields.</p>

<p>From what I remember, we now have HTML being escaped through a dependency added to pom.xml. There should be a way to tell Jenkins/this dependency that we know what we are doing in certain cases... or which tags we want.</p>

<p>But I haven't tested it yet. After you create the new JIRA issue for that, if you could include a simple config.xml, I'll then use it for testing &amp; patching the code.</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Hi Ioannis!</p>

<p>Thanks for testing it so quickly. Perhaps we can work on a patch for this one too? First I will investigate if we can enable the HTML for those fields.</p>

<p>From what I remember, we now have HTML being escaped through a dependency added to pom.xml. There should be a way to tell Jenkins/this dependency that we know what we are doing in certain cases... or which tags we want.</p>

<p>But I haven't tested it yet. After you create the new JIRA issue for that, if you could include a simple config.xml, I'll then use it for testing & patching the code.</p>

<p>Bruno</p>
